### PR TITLE
v4.1

### PR DESCRIPTION
**Pull request by binaryoverload**
## 4.1 release
FlareBot v4.1 - The riot change

## Added
Command - usage - example
* tempban - `_tempban <user> <duration> [reason]` - `_tempban weeryan17 1d Many spellinz izzues`
* tempmute - `_tempmute <user> <duration> [reason]` - `_tempmute Charlizard 12h Too much rioting`
* Reminders! (Beta) - `_remind <duration> <reminder>` - `_r 1d Check feedback on remind command`
* Dynamic Usage System - Ooo! It will now show the usage of the actual arg (fancy stuffs!) 
* An "extra info" for commands for display under usage (only used by welcome as of now)
* More aliases!!
  * Selfassign now has `iam`, `sa` and `giverole`
  * `quit` is now an alias of leave
* A safeguard method to prevent double spaces from messing up commands. Any amount of spaces are now replaced with the one. So you can now do them `_ban Walshy` commands with 100 spaces before Walshy's name!! YAY!
* We have heavily expanded on modlog and allow you to log just about anything!! Huge thanks @weeryan17 for his work on this.
  * View the modlog command for all the info - `_usage modlog`

## Changed
* Sharding logic again to better suit the new way with JDA
* Shardinfo command now shows some more info - websocket ping and VCs
* Some permissions stuff - added `_permissions reset` which will reset all permissions and groups also added `_permissions restoredefault` which is good when we add new perms BUT will re-add removed perms. **Use with caution.**
* Change the purge command - You can now purge users! New usage `_purge <amount/user> [amount]` example: `_purge BinaryOverload 50` or `_purge 50`
* Purge will log to the modlog! It will show who was purged (if applicable), the number of messages and who purged.
* Added a `flarebot.playlist.clear` permission **Highly requested** - This is a default permission. Remove it with `_permissions group Default remove flarebot.playlist.clear`
* Selfassign will now accept roles name to add and remove! New usage: `_selfassign add/remove <rolename|id>`
* Reports creations and report changes now go to the modlog! 
* Refactor *all* usages to work with the new system
* The bot no longer deafens itself - This didn't do anything and some users got confused by it.
* Changed categories of a few commands
  * `convert` and `currency` now have their own category called "currency"
  * `join` and `leave` are now in the "music" category
  * Added an `avatar` command which is in the "random" category
  * Moved `poll` and `tags` along with the new `remind` commands to a new category called "useful"
* **Search no longer functions** - You have to use the new command `play` which no longer resumes on 0 args.

## Fixed
* Internal NPEs
* Issue where `ban` would throw an error if the user didn't exist.
* Bug where you'd be DM'd since we can give other bots roles.
* Ryan's Code in permissions being repetitive and unreliable
* Usage now uses the embed description to stop char overflows
* Bug where 5 reactions would pin a message, not just pushpins.
* The fix command not saying to do `fix confirm`

## Beta
* `_currency` and `_convert` are now out of beta! Make sure to find any bugs and report them! :)
* You can now edit tags using `_tag edit <tag_name> <tag_message>` as opposed to removing and re-adding!

### CODE REVIEW

File: src/main/java/stream/flarebot/flarebot/FlareBot.java [hunk 1]
```diff
@@ -222,6 +199,9 @@ public static void main(String[] args) throws Exception {
         String tkn = config.getString("bot.token").get();
 
         new CassandraController().init(config);
+
+        /*if (config.getString("bot.pasteAccessKey").isPresent())
```
Comment (weeryan17) [inline src/main/java/stream/flarebot/flarebot/FlareBot.java]: why is this commented out?

Comment (WalshyDev) [reply src/main/java/stream/flarebot/flarebot/FlareBot.java]: All of that is commented out to make use of a global config variable instead. Much cleaner and generally more efficient code.

**Comment by weeryan17** [comment]
We should probably rename QuitCommand to ShutdownCommand or something else like that.

**Comment by WalshyDev** [comment]
@weeryan17 this isn't really a needed change anymore, as can be seen [here](https://github.com/FlareBot/FlareBot/blob/1cc8497bca4941a36052c3ed91287b967f03d971/src/main/java/stream/flarebot/flarebot/FlareBot.java#L803-L814) the command logic has changed to find an actual command that isn't a secret if they aren't a dev. With this LeaveCommand has the alias of "quit" which now works for none-developers while we still maintain our normal command which we are used to.

**Comment by weeryan17** [comment]
ok

**Comment by WalshyDev** [comment]
No one saw that. Should we wait for shard-manager before merge?

**Comment by binaryoverload** [comment]
a) How far off is shard manager?
b) Why the fuck do we have merge conflicts?

**Comment by binaryoverload** [comment]
Oh yeah that's true xD Someone merge master into hacktoberfest plskthnx

**Comment by weeryan17** [comment]
You're a dev do it yourself :P

**Comment by WalshyDev** [comment]
Shard manager is close, we can also just work off the git hash it's stable and up-to-date

**Comment by weeryan17** [comment]
Are we planning on having a page system for 4.1?

**Comment by WalshyDev** [comment]
Depends on if it can be done on time, feel free to finish it

**Comment by weeryan17** [comment]
I have modlog stuff to finish. I can finish that pretty quickly over the break i have coming up i might be able to do pages

**Comment by weeryan17** [comment]
ummm @binaryoverload your permissions command stuff is broken. you can no longer create roles

**Comment by WalshyDev** [comment]
The infraction system that this PR was waiting on will actually be moved to 4.2
Reason: Like the reports (in 4.2) they will have their own table and will be loaded/cached when needed. It would be best to wait until a cleanup is done/in-progress to then work on logic for this kind of thing. We should have a proper system for this so that we can support a lot more load-on-demand in the future with ease.

A cleanup commit is incoming then this will be merged.

**Pull request merged**